### PR TITLE
Do not add other route when catchall route is built

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
When there are multi HTTPRoute.Match set, if one is catch all, then ignore other match conditions, so no need to build envoy route for them.

VirtualHost.Routes will be matched in order.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[x] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

File: pilot/pkg/networking/core/v1alpha3/route/route.go [hunk 1]
```diff
@@ -271,29 +271,30 @@ func BuildHTTPRoutesForVirtualService(
...
-			for _, match := range http.Match {
-				if r := translateRoute(push, node, http, match, listenPort, virtualService, serviceRegistry, gatewayNames); r != nil {
+			return out, nil
+		}
+
+		tmp := []*route.Route{}
+		for _, match := range http.Match {
+			if r := translateRoute(push, node, http, match, listenPort, virtualService, serviceRegistry, gatewayNames); r != nil {
+				tmp = append(tmp, r)
+				// This is a catch all path. Routes are matched in order, so we will never go beyond this match
```
Comment (howardjohn) [inline pilot/pkg/networking/core/v1alpha3/route/route.go:287]: I vaguely recall we used to do this and reverted it for some reason.. let me see if I can dig up the history

Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/route/route.go:287]: https://github.com/istio/istio/pull/27449 was what I was thinking of

Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/route/route.go:287]: HTTPRoute.Match
	// Match conditions to be satisfied for the rule to be
	// activated. All conditions inside a single match block have AND
	// semantics, while the list of match blocks have OR semantics. The rule
	// is matched if any one of the match blocks succeed.

This is the api doc, it says it matches when anyone condition matches. If there is a match all, all other routes are redundant

Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/route/route.go:287]: @howardjohn After a close look at the issue, it is different, with this pr it does not remove the `provider-global-default` catch all route. There is ut for this. And i am testing manually in my env

Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/route/route.go:287]: Verified it does not remove the catchall `provider-global-default`
```
[
    {
        "name": "8000",
        "virtualHosts": [
            {
                "name": "allow_any",
                "domains": [
                    "*"
                ],
                "routes": [
                    {
                        "name": "allow_any",
                        "match": {
                            "prefix": "/"
                        },
                        "route": {
                            "cluster": "PassthroughCluster",
                            "timeout": "0s",
                            "maxGrpcTimeout": "0s"
                        }
                    }
                ],
                "includeRequestAttemptCount": true
            },
            {
                "name": "httpbin.default.svc.cluster.local:8000",
                "domains": [
                    "httpbin.default.svc.cluster.local",
                    "httpbin.default.svc.cluster.local:8000",
                    "httpbin",
                    "httpbin:8000",
                    "httpbin.default.svc",
                    "httpbin.default.svc:8000",
                    "httpbin.default",
                    "httpbin.default:8000",
                    "10.96.196.41",
                    "10.96.196.41:8000"
                ],
                "routes": [
                    {
                        "name": "GET-/",
                        "match": {
                            "prefix": "/",
                            "caseSensitive": true,
                            "headers": [
                                {
                                    "name": ":method",
                                    "exactMatch": "GET"
                                }
                            ]
                        },
                        "route": {
                            "cluster": "outbound|8000||httpbin.default.svc.cluster.local",
                            "timeout": "4s",
                            "retryPolicy": {
                                "retryOn": "gateway-error,connect-failure,retriable-status-codes",
                                "numRetries": 3,
                                "perTryTimeout": "1s",
                                "retryHostPredicate": [
                                    {
                                        "name": "envoy.retry_host_predicates.previous_hosts"
                                    }
                                ],
                                "hostSelectionRetryMaxAttempts": "5"
                            },
                            "maxGrpcTimeout": "4s"
                        },
                        "metadata": {
                            "filterMetadata": {
                                "istio": {
                                    "config": "/apis/networking.istio.io/v1alpha3/namespaces/default/virtual-service/bar"
                                }
                            }
                        },
                        "decorator": {
                            "operation": "httpbin.default.svc.cluster.local:8000/*"
                        },
                        "requestHeadersToAdd": [
                            {
                                "header": {
                                    "key": "x-envoy-retriable-status-codes",
                                    "value": "408,423"
                                },
                                "append": true
                            }
                        ]
                    },
                    {
                        "name": "provider-global-default",
                        "match": {
                            "prefix": "/"
                        },
                        "route": {
                            "cluster": "outbound|8000||httpbin.default.svc.cluster.local",
                            "timeout": "0s",
                            "retryPolicy": {
                                "retryOn": "connect-failure,refused-stream,unavailable,cancelled,retriable-status-codes",
                                "numRetries": 2,
                                "retryHostPredicate": [
                                    {
                                        "name": "envoy.retry_host_predicates.previous_hosts"
                                    }
                                ],
                                "hostSelectionRetryMaxAttempts": "5",
                                "retriableStatusCodes": [
                                    503
                                ]
                            },
                            "maxGrpcTimeout": "0s"
                        },
                        "metadata": {
                            "filterMetadata": {
                                "istio": {
                                    "config": "/apis/networking.istio.io/v1alpha3/namespaces/default/virtual-service/bar"
                                }
                            }
                        },
                        "decorator": {
                            "operation": "httpbin.default.svc.cluster.local:8000/*"
                        }
                    }
                ],
                "includeRequestAttemptCount": true
            }
        ],
        "validateClusters": false
    }
]

```

File: pilot/pkg/networking/core/v1alpha3/route/route.go [hunk 2]
```diff
@@ -271,29 +271,30 @@ func BuildHTTPRoutesForVirtualService(
...
-			if r := translateRoute(push, node, http, nil, listenPort, virtualService, serviceRegistry, gatewayNames); r != nil {
+			r := translateRoute(push, node, http, nil, listenPort, virtualService, serviceRegistry, gatewayNames)
+			if r != nil {
 				out = append(out, r)
 			}
-			catchall = true
-		} else {
-			for _, match := range http.Match {
-				if r := translateRoute(push, node, http, match, listenPort, virtualService, serviceRegistry, gatewayNames); r != nil {
+			return out, nil
```
Comment (howardjohn) [inline pilot/pkg/networking/core/v1alpha3/route/route.go:280]: the old code would return an error if translateRoute returns nil. is the behavior intentionally changed here? (L300 of old code)

File: pilot/pkg/networking/core/v1alpha3/route/route.go (hunk 1 above)
Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/route/route.go:287]: Thanks for looking into it!

**Review by howardjohn — CHANGES_REQUESTED** [review]
I think this is very close to being a safe optimization. However, my concern is the impact on tracing.

Consider a route like:
```
  - match:
    - name: foo
      uri:
        prefix: /foo
    - uri:
        prefix: /
```

This is actually NOT exactly equivalent to 
```
  - match:
    - uri:
        prefix: /
```

Because we won't get a segment like
```
    - decorator:
        operation: echo.default.svc.cluster.local:80/foo*
```

in the envoy configuration.

One thing I noticed - we do already have a webhook warning for this case. So if a user does accidentally set this, they will be alerted and can change the VS themselves to optimize it?

WDYT?

File: pilot/pkg/networking/core/v1alpha3/route/route.go (hunk 2 above)
Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/route/route.go:280]: It is to short circuit it, here it always return a catchall route?

**Comment by hzxuzhonghu** [comment]
Yes. the above case equals to: 
```
- match:
    - uri:
        prefix: /
```

We can improve the validation

**Comment by howardjohn** [comment]
> Yes. the above case equals to:
> 
> ```
> - match:
>     - uri:
>         prefix: /
> ```
> 
> We can improve the validation

That was my point. These are **not** equal. They generate different traces, because the traces include the URL match

**Comment by hzxuzhonghu** [comment]
Yes, the trace could be different, but does that make much sense? In real case, if user allows traffic matching /a or /b or anything, routed to the same destination,  they don't care about what span name it is. There is not stats inferred from this.
 User mostly care about original and destination service, and related attributes from traces.

**Comment by howardjohn** [comment]
I am not sure it's safe to assume that users don't care about it, I could easily see folks depending on this. It would be good to get some more feedback from @istio/wg-policies-and-telemetry-maintainers .

The biggest complaints about Istio we are seeing is subtle breaking changes like this so I am a hesitant to make then where possible

**Comment by mandarjog** [comment]
Do we allow a way to indicate metadata, decorator or typed-filter-config ?

we should not apply the optimization of any of these things are present. For example a filter may use the some route metadata to make a different decision even though the destination is the same.

**Comment by hzxuzhonghu** [comment]
@mandarjog From envoy doc, decorator only influences the trace span name. 

```
// The operation name associated with the request matched to this route. If tracing is
	// enabled, this information will be used as the span name reported for this request.
```

It has nothing todo with metadata. Below is how the span name is generated, I think we are too concerned about such small changes.  Honestly, there is no change that breaks compatibility can match mixer deprecation.

```
func getRouteOperation(in *route.Route, vsName string, port int) string {
	path := "/*"
	m := in.GetMatch()
	ps := m.GetPathSpecifier()
	if ps != nil {
		switch ps.(type) {
		case *route.RouteMatch_Prefix:
			path = m.GetPrefix() + "*"
		case *route.RouteMatch_Path:
			path = m.GetPath()
		case *route.RouteMatch_SafeRegex:
			path = m.GetSafeRegex().GetRegex()
		}
	}
```

**Pull request closed without merging**